### DISCUSSION TITLE
[Feature] Tree structured album folders

### DISCUSSION THREAD

**Discussion by waclaw66** [immich-app/immich#5078] [Feature Request]
### The feature

When someone has a high number of albums the browsing is clumsy and slow.
As I heard from my friends and photographers, everyone is looking for a photo storage service that could provide tree stuctured albums.
I would suggest to introduce a new functionality to structure albums in folders. Albums then would act like files.
There would be a new action button "Create folder" that creates a new folder. Folder context menu has items "Delete folder" (deletes only empty folder) and "Move to" (moves folder to any other folder). Folders would be shown (ordered by name) before albums in the current tree node. Album context menu has a new item "Move to" (moves album to any another folder).
It wouldn't change anything for those who do not want to use folders, but it would be a huge improvement for the majority of others.

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by bo0tzz** [comment]
Duplicate of #2073

**Reply by waclaw66** [reply]
Not really the duplicate. Nested albums are quite complicated, this is just folder layer above albums. It's quite simple feature.

**Comment by briandipalma** [comment]
There seem to be at least two issues with this approach.

1. It clashes with the Storage Template feature. A filepath can either be controlled by the Storage Template or this folder album based approach.
2. A picture can only be in one album at a time.

I know I sound like a broken record but hierarchical tags are an excellent solution for this problem and a great replacement for albums. I have top level tags like `2012`, `2019`, `2014` etc, inside I have subtags like `October`, `January` then inside those I have tags such as `Spain`, `Italy`, `London` etc and inside those tags such as `London Zoo`, `Barcelona`, `Ostia Lido` etc. The level of nesting is up to me, each time I click on a sublevel of nesting I see fewer and fewer pictures.

The tags are saved to the picture file metadata. So I can nuke my Immich instance/photo app database and reload the files and voila all my tags are regenerated on metadata reading. I can use other software to read and update the tags and when Immich re-reads the data it sees the changes. Pictures can have multiple tags applied to them.

This is the right way to deal with albums and nested albums. Albums as Google created them was a way to make GPhotos "sticky", all that hardwork sorting your pictures is only available via GPhotos. Metadata tags are the open and portable approach to GPhotos albums.

**Reply by dkbarn** [reply]
> hierarchical tags are an excellent solution for this problem and a great replacement for albums.

Are they though? Can you share the contents of a tag with other people like you can with an album?

**Reply by briandipalma** [reply]
There's no reason you couldn't. It needs to be implemented but there's nothing fundamentally blocking that functionality.

**Reply by RamblinWreck77** [reply]
>I know I sound like a broken record but hierarchical tags are an excellent solution for this problem

A folder structure from windows 95 can still be applicable/work in 2025. Any "tag" system is bespoke and will break the moment you move to a new system.

**Reply by briandipalma** [reply]
Tags aren't bespoke to Immich, I create and edit my tags in Digikam and they show up and work fine in Immich, they are metadata tag in the picture file or a sidecar file. It's pretty standard in photo management software. You are thinking of Immich albums which are tied to Immich.

**Comment by waclaw66** [comment]
@briandipalma your arguments are not valid. This is about upper layer to albums only, it doesn't interfere storage or album-asset relations. It's just virtual album management. The only change to the current db schema for albums is parent folder id.

|albums|
|-|
|...|
|folderId: uuid[null]|

|folders|
|-|
|id: uuid|
|parentId: uuid[null]|
|name: varchar|

**Reply by briandipalma** [reply]
So it's not an actual folder with actual files in it? It's all virtual? Well in which case why bother? This is a more limited tags implementation.

**Reply by waclaw66** [reply]
Yes it's all virtual. The purpose is for better albums managing. Instead of having hunderts of albums in a single list you can have tree folder stucture of albums. It doesn't care of tags.

**Reply by briandipalma** [reply]
Hierarchical tags will give you the tree structure you are asking for in a way that darktable, digikam, exiftool and many other picture management tools already support and in a way that can be stored within the file itself so you won't lose it.

**Reply by waclaw66** [reply]
That's possible, but this feature topic is not about tags. Feel free to start a new feature topic.

**Comment by alextran1502** [comment]
I like this approach. Can you help me understand the list of pros and cons of "folder" vs. "nested album"

**Reply by briandipalma** [reply]
Nested albums using hierarchical tags stored in the picture's metadata. Open standard used by multiple other photo organizing programs. There is no need for albums as GPhotos thinks of them once you have this. Quick filter through the photos to delete poor ones, create your hierarchical tags, tag the remaining photos with keyboard shortcuts. Job done. Open pictures in digikam or darktable and they also read the tag metadata so it's portable.

![image](https://github.com/immich-app/immich/assets/1597820/9f72ddf4-9442-4807-98fe-b3989e98936f)

**Reply by waclaw66** [reply]
Pretty well described...
https://www.phind.com/search?cache=dye9lxx7xsrbjs5vk56aae6x

Tree structure is achievable with both, but hierarchical tags are way more complex.

**Reply by MD77MD** [reply]
@alextran1502 
> I like this approach. Can you help me understand the list of pros and cons of "folder" vs. "nested album"

I think tages and nested albums are much better and this is why

say I have personal albums, family albums and friends albums:
* in folder system there is no way to have a common picture in all folder unless you duplicate it
* however in (nested tags/albums) system you can have such picture in multiple album with multiple tags virtual without the need to duplicate the picture 

non the less, I think what Synology had was correct choose by implanting both system as everyones needs are different 
**I think the perfect balance would have us think of the folder system as a initial basic step to organize photos when imported then user would use the  (nested tags/albums) system to actually do the organization onword.**

**Comment by hgpuke** [comment]
+1 for this. I already have my photos (in an external library) organized this way, and it is a pity that Immich currently cannot take advantage of this structure. I used to have this feature in Synologys Photostation, but now I've stopped using Synology and run my own server and need this function badly. In the mean time, I run Photoview in parallel to Immich. Photoview does _exactly_ this.

**Comment by seitzbg** [comment]
+1 for this from me as well. Will 100% purchase a license if this will be implemented soon.

**Comment by MD77MD** [comment]
@alextran1502 
`I like this approach. Can you help me understand the list of pros and cons of "folder" vs. "nested album"`

I think tages and nested albums are much better and this is why

say I have personal albums, family albums and friends albums:

in folder system there is no way to have a common picture in all folder unless you duplicate it
however in (nested tags/albums) system you can have such picture in multiple album with multiple tags virtual without the need to duplicate the picture
non the less, I think what Synology had was correct choose by implanting both system as everyones needs are different
I think the perfect balance would have us think of the folder system as a initial basic step to organize photos when imported then user would use the (nested tags/albums) system to actually do the organization onword.

**Comment by MD77MD** [comment]
@alextran1502 

I think tages and nested albums are much better

say I have personal albums, family albums and friends albums:
* in folder system there is no way to have a common picture in all folder unless you duplicate it
* however in (nested tags/albums) system you can have such picture in multiple album with multiple tags virtual without the need to duplicate the picture 

non the less, I think what Synology had was correct choose by implanting both system as everyones needs are different 
**I think the perfect balance would have us think of the folder system as a initial basic step to organize photos when imported then user would use the  (nested tags/albums) system to actually do the organization onword.**

**Comment by Mannshoch** [comment]
I would suggest a virtual folder concept with WebDAV. Let the Server storage be as it is. But allow to set up some virtual WebDAV endpoint that:

- contain selected photos 
- show a structure setup for this endpoint. 
  - similar to the library
  - (more creative idea:)
  allow set several tags that act as virtual folder on certain level.
  yes, that way photos could appear several times in the structure
 creating a folder and copy a picture in it would mean a picture is tagged with that.
  - removing a photo could mean either move it to a bin or remove it from the selected photos of this structure or add a tag (deleted) so you could recheck it in immich.

**Comment by Baxter117** [comment]
I think this is a necessary feature to use the app. otherwise it does not replace a folder structure where i sort my photos in it.
and when I have to sort all data on my PC manually in folders, I have no benefit from using immich.
Also my parentes cannot move to immich, because they need their folder structured photos, and want so sort new photos into this structure.

**Comment by waclaw66** [comment]
https://opennoodle.de/features/space-album-folders/
That's exactly what I described above. I would be happy to see this feature also within Immich.

**Reply by briandipalma** [reply]
Tags and nested tags

**Comment by moerphie** [comment]
I'm wondering why such a basic feature has been discussed for years but hasn't been implemented.
For anyone who wants to switch from iOS Photos to Immich, this will be a deal-breaker. There, you can organize everything into folders and albums.
